### ISSUE TITLE
Access to /dev/ttyUSB0

### ISSUE THREAD

**Issue by 21stce** [NixOS/nixpkgs#162542]
After inserting a device /dev/ttyUSB0 becomes available, but how do I make it such that the permissions are set such that a user or set of users has access to it? I'd expect a suitable udev rule would accomplish this. 

In the context of Arduino, it would be useful if there were a NixOS arduino module, which would use the above non-existing mechanism. 

https://stackoverflow.com/questions/59925105/nix-package-requires-sudo-to-access-usb-device provides further context, but should not be required to understand the above.

**Comment by jansol** [comment]
What would such a module do? To access `/dev/ttyUSB0` your user needs to be in the "dialout" group (see `users.users.<username>.extraGroups`. If, for whatever reason the group is not "dialout" you can find out which one it is with `ls -al /dev/ttyUSB0`.

**Comment by 21stce** [comment]
https://nixos.org/manual/nixos/stable/ does not even contain the word `dialout`. 

`man configuration.nix` also does not specify what `dialout` is supposed to do. 

Just because other operating systems might do that, doesn't mean NixOS also does that, unless the manual specifies that "it does the same as distro X".

**Comment by jansol** [comment]
There is nothing NixOS-specific about this. It is simply what [udev does by default](https://github.com/systemd/systemd/blob/da60d91c9908b33fa6baa10ef80a7adc2312e615/rules.d/50-udev-default.rules.in#L25).

Not sure if the manual really needs a chapter on arduino/serial consoles, but it would definitely be a useful addition to the [community wiki](https://nixos.wiki/). That said, if you do have a good idea for what such a nixos module / chapter in the manual would contain I'm all ears.

**Comment by 21stce** [comment]
We could choose not to invent a square wheel: https://docs.freebsd.org/doc/6.3-RELEASE/usr/share/doc/handbook/serialconsole-setup.html.

Ideally, when the Arduino IDE is started, it could just say "Please ask your administrator to add the following text to your configuration.nix and it will work. ".

**Comment by rnhmjoj** [comment]
> Ideally, when the Arduino IDE is started, it could just say 

> Just because other operating systems might do that, doesn't mean NixOS also does that, unless the manual specifies that "it does the same as distro X".

I think you're fundamentally misunderstanding how NixOS and it's development work.
You're expecting there's people having full understanding and control over the entire operating system and applications; that may be the case for some proprietary OSs or BSDs, but it's certainly not the case of NixOS.

NixOS, being a GNU/Linux distribution, is a curated collection of software components (independently developed by many different groups) that has merely been configured to work more or less well together.
From a user perspective this is not great: sometimes it can be hard to even find which component is at fault when something goes wrong, but that's the reality of it.

> https://nixos.org/manual/nixos/stable/ does not even contain the word dialout.

You can't reasonably ask for the NixOS manual to contain an explanation of the >80000 software packages
distributed. NixOS has its peculiarities (many) and the manual mainly covers that, along with other common tasks and how specific software interacts with each other, but it's not meant to be a comprehensive guide to Unix systems administration.

**Comment by 21stce** [comment]
> You're expecting there's people having full understanding and control over the entire operating system and applications; that may be the case for some proprietary OSs or BSDs, but it's certainly not the case of NixOS.

How do you mean "control"? Access rights? Or the ability to change literally everything across the stack? As for the first, I have root (and many developers have). As for the last, NixOS is uniquely suited to provide extreme customization and integration and a great out of the box experience. 

In short, I have no idea what you are trying to communicate. 

>  but it's not meant to be a comprehensive guide to Unix systems administration.

Clearly, whoever started this had an idea in mind for how groups should work, right? As such, that person could just write down "I was a Debian user and I have modeled everything that isn't documented like Debian". That seems like a perfectly valid way to resolve the lack of documentation. If you are saying NixOS is just a huge mess, then I suppose the question is: why and do you want to keep it like that?

**Comment by rnhmjoj** [comment]
> I have root (and many developers have). As for the last, NixOS is uniquely suited to provide extreme customization and integration and a great out of the box experience.

Yes, but to some extents: for example I don't think NixOS patching the Arduino IDE to add a warning is a sensible idea.
If it's beneficial to all users the changes should be done in the original project, not NixOS. So, report that to the Arduino people.

In general these sort of changes requires maintaining a significant amount of code downstream, i.e. as patches outside of the original project. This is generally dreaded upon by maintainers because it slows down the process of upgrading a package (when a conflicts happens the patches need to be rebased manually) and divergences in the behavior of the programs can cause bugs and security issues. 

NixOS already does some extensive modifications to the compiler toolchain and systemd, but these are absolutely essential for the goal of the project (purity, reproducibility, etc) and sometimes can't be upstreamed.

> Clearly, whoever started this had an idea in mind for how groups should work, right? As such, that person could just write down "I was a Debian user and I have modeled everything that isn't documented like Debian".

No, I don't a single person in NixOS (or Debian) sat down and decided how groups should work. The default system groups are inherited by various components like systemd, CUPS, X.org etc. In this case the `dialout` group is something udev assigns to certain devices in the default rules, so it was simply carried over (0f1a368d75d2151b1ccc6da0cbd5e0f7510542e5).

> That seems like a perfectly valid way to resolve the lack of documentation.

Uhm, I'm so sure: if we just refer to what distro X, we should also make sure it continues to act as NixOS does. At this point it's easier to document what NixOS does.

> If you are saying NixOS is just a huge mess, then I suppose the question is: why and do you want to keep it like that?

No, I would say it's only a moderate mess ;)
The reason for why a lot of things are broken, missing or in an undesirable state is usually due to the lack of people volunteering their time to make them right. So, If you find this important I'd suggest you make the changes you'd like to see.

**Comment by 7c6f434c** [comment]
> Clearly, whoever started this had an idea in mind for how groups should work, right? As such, that person could just write down "I was a Debian user and I have modeled everything that isn't documented like Debian". That seems like a perfectly valid way to resolve the lack of documentation. 

No. Devices are added by `udev`. Whenever it is reasonable-ish, we just keep `udev` defaults. In general, in Nix* we keep upstream defaults when it doesn't create Nix*-specific problems, we don't bother whether Debian changes these defaults or not unless we have a complicated trade-off discussion and looking up arguments made by major distributions for their choces.

> If you are saying NixOS is just a huge mess, then I suppose the question is: why and do you want to keep it like that?

NixOS is implicitly aimed at people who know enough of the GNU/Linux defaults and have detailed enough preferences to actually have a chance of benefitting from it. I want at least Nixpkgs to have reaonably close to upstream behaviours, because a) we already lack the throughput for doing update as well as people would prefer, b) I don;t want distro-specific weird behaviours.

Yes, if you go and write a Wiki page and put «as of March 2022» in the text, it will be a net positive. NixOS manual entry… I am not even sure about that. NixOS module for device access, probably useful.

**Comment by jansol** [comment]
> Ideally, when the Arduino IDE is started, it could just say "Please ask your administrator to add the following text to your configuration.nix and it will work. ".

Cargo-culting like "Paste the following into your config" is universally bad advice that results in unnecessary cruft and sooner or later some really bizarre breakage.

I do agree that Arduino IDE could remind people to make sure their user is part of the relevant groups, however that is best left to upstream since they will have the most knowledgeable people for this. They can even keep track of which groups are needed for specific devices/actions and show a "ask your administrator to add your user to the X group(s)" when you are attempting stuff where that is relevant, if your user is not already in said group(s). That way it'll benefit all their users regardless of distro.

"Your administrator" should know how to do basic things like modifying user groups on whatever system it is you are using, so no need to specify that in a distro-specific way in the application. And if "your administrator" doesn't know how to do it yet, *that* is a thing that belongs in the distro-specific manual.
